sql/rls: exempt foreign key propagation from RLS enforcement

spilchen opened the pull request:
Previously, RLS policies were applied during foreign key propagation, triggered by ON DELETE or ON UPDATE actions specified in foreign key definitions.  This behavior was incorrect because foreign key maintenance operations must not be blocked by RLS to ensure data integrity. Postgres exempts such operations, and we now match that behavior.

Fixes #145333

Epic: CRDB-11724
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145341)
<!-- Reviewable:end -->

mgartner reviewed (approved):
LGTM. Some optbuilder tests would be nice, including tests for `ON DELETE CASCADE` and `ON UPDATE SET NULL`.

spilchen committed:
sql/rls: exempt foreign key propagation from RLS enforcement

Previously, RLS policies were applied during foreign key propagation,
triggered by ON DELETE or ON UPDATE actions specified in foreign key
definitions.  This behavior was incorrect because foreign key
maintenance operations must not be blocked by RLS to ensure data
integrity. Postgres exempts such operations, and we now match that
behavior.

Fixes #145333

Epic: CRDB-11724
Release note: none

spilchen commented:
TFTR!

bors r+
